IPC service: RPMsg backend: initialize shared memory to zero

kapi-no opened the pull request:
Added a code for initializing shared memory to zero. This operation
normalizes the memory state so that the IPC service is no longer
prone to reading status bits from the previous reset session.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no committed:
ipc: backends: rpmsg: initialize shared memory to zero

Added a code for initializing shared memory to zero. This operation
normalizes the memory state so that the IPC service is no longer
prone to reading status bits from the previous reset session.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

hubertmis commented on subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:
```diff
@@ -494,4 +496,26 @@ static int backend_init(const struct device *instance)
+	     backend_config < backend_configs + ARRAY_SIZE(backend_configs);
+	     backend_config++) {
+		if (backend_config->role == ROLE_HOST) {
+			memset((void *) backend_config->shm_addr, 0, backend_config->shm_size);
```
nit: should be enough to set 4 bytes:

```suggestion
			memset((void *) backend_config->shm_addr, 0, VDEV_STATUS_SIZE);
```

carlocaione replied:
`z_early_memset`

carlocaione commented on subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:
```diff
@@ -494,4 +496,26 @@ static int backend_init(const struct device *instance)
+		DT_INST_FOREACH_STATUS_OKAY(BACKEND_CONFIG_DEFINE)
+	};
+
+	for (backend_config = backend_configs;
+	     backend_config < backend_configs + ARRAY_SIZE(backend_configs);
+	     backend_config++) {
```
This is unnecessarily complicated. `DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT)` gives you the number of instances, and then you can simply get address and size with `DT_REG_{SIZE,ADDR}(DT_INST_PHANDLE(i, memory_region))`

carlocaione commented on subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:
```diff
@@ -494,4 +496,26 @@ static int backend_init(const struct device *instance)
+	     backend_config < backend_configs + ARRAY_SIZE(backend_configs);
+	     backend_config++) {
+		if (backend_config->role == ROLE_HOST) {
+			memset((void *) backend_config->shm_addr, 0, backend_config->shm_size);
```
There is no need to `memset` the whole region, the status region should be enough.

kapi-no replied:
fixed

carlocaione reviewed (changes requested):
I'm still not entirely sold on the opportunity to have this in the backend. I need to think about this but the other possibility is to have this as part of the SoC specific initialization code instead of having this here.

carlocaione commented:
@arnopo do you have any suggestion for this (maybe you saw the same issue on other platforms).

hubertmis commented:
@arnopo, I believe I can answer your questions.

> * When you say "reading status bits from the previous reset session", do you means that you shared memory is in a non volatile memory, that is preserved on a reset?

It's about "reset" definition. In some systems a core can be reset, while RAM content (volatile in its nature) is preserved. IPC should correctly handle such scenario.

> Is a reset reason is a crash of the "host" core?

I think multiple reasons can be considered here. One option is a crash, another one is soft reset request on the host side (an application request).

> What happen when only the remote core reset ( on crash)?

I believe host would need to:
1. detect this condition (assuming the remote core does not start automatically on its reset)
2. clear shared memory
3. reinitialize host's OpenAMP instance
4. start remote core

But this procedure is out of scope of this patch.

carlocaione commented on subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:
```diff
@@ -494,4 +496,26 @@ static int backend_init(const struct device *instance)
+
+static int shared_memory_prepare(const struct device *arg)
+{
+	const struct backend_config_t *backend_config;
+	const struct backend_config_t backend_configs[] = {
+		DT_INST_FOREACH_STATUS_OKAY(BACKEND_CONFIG_DEFINE)
+	};
+
+	for (backend_config = backend_configs;
+	     backend_config < backend_configs + ARRAY_SIZE(backend_configs);
+	     backend_config++) {
+		if (backend_config->role == ROLE_HOST) {
+			memset((void *) backend_config->shm_addr, 0, VDEV_STATUS_SIZE);
+		}
+	}
```
I think we could pull this off with a way smaller change. Maybe something like:
```
#define BACKEND_CONFIG_INIT(n) config[n] = &backend_config_##n;
static int shared_memory_prepare(const struct device *arg)
{
       static struct backend_config_t
               *config[DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT)];

       DT_INST_FOREACH_STATUS_OKAY(BACKEND_CONFIG_INIT);

       for (int i = 0; DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT); i++) {
               ...
       }
}
SYS_INIT(shared_memory_prepare, PRE_KERNEL_1, 1);
```
without resorting to `BACKEND_CONFIG_POPULATE`.

carlocaione replied:
@kapi-no can you try using this solution? I guess the changed LOC will be smaller.

kapi-no replied:
@carlocaione, sure, I can change it anytime

carlocaione commented on subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:
```diff
@@ -495,3 +495,25 @@ static int backend_init(const struct device *instance)
+
+#define BACKEND_CONFIG_INIT(n) config[n] = &backend_config_##n;
+
+#if CONFIG_IPC_SERVICE_BACKEND_RPMSG_SHMEM_RESET
```
`#ifdef` or `#if defined()`

kapi-no replied:
This will always be defined but depending on its value 0 or 1 it is considered enabled or disabled. That's why I used #if macro

carlocaione replied:
boolean symbols not set or set to `n` do not end up in `autoconf.h` so they are not actually defined (they are not set to 0). In general Zephyr devs prefer to use '#ifdef` when both are fine.

carlocaione commented on subsys/ipc/ipc_service/backends/Kconfig.rpmsg:
```diff
@@ -12,4 +12,13 @@ config IPC_SERVICE_BACKEND_RPMSG_WQ_STACK_SIZE
 
+config IPC_SERVICE_BACKEND_RPMSG_SHMEM_RESET
+	bool "Reset shared memory state"
+	default y
```
uhm, this should not usually be required so let's stick to `n`. "broken" platforms can set this to `y` instead.

kapi-no replied:
I set it up as you suggested

carlocaione commented on subsys/ipc/ipc_service/backends/Kconfig.rpmsg:
```diff
@@ -12,4 +12,13 @@ config IPC_SERVICE_BACKEND_RPMSG_WQ_STACK_SIZE
+	bool "Reset shared memory state"
+	default y
+	help
+	  This configuration normalizes the memory state so that the IPC
+	  service is no longer prone to reading status bits from the
+	  previous reset session. The status field of the shared memory
+	  is set to zero.
```
what about something along the lines of:

`Some platforms retain the memory content upon reset. This is problematic because the backend expects a zero-ed memory to be able to correctly setup instances and endpoints at init time. When this parameter is set to 'y' the status region of the shared memory is reset on kernel initialization.`

or something like this.

kapi-no replied:
Rephrased the prompt according to your suggestion

carlocaione commented on subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:
```diff
@@ -495,3 +495,25 @@ static int backend_init(const struct device *instance)
+}
+
+SYS_INIT(shared_memory_prepare, PRE_KERNEL_1, 1);
+#endif /* defined(CONFIG_IPC_SERVICE_BACKEND_RPMSG_SHMEM_RESET) */
```
```suggestion
#endif /* CONFIG_IPC_SERVICE_BACKEND_RPMSG_SHMEM_RESET */
```

kapi-no replied:
fixed

carlocaione reviewed (dismissed):
Besides the nit LGTM

hubertmis commented on boards/arm/bl5340_dvk/Kconfig:
```diff
@@ -13,6 +13,13 @@ config MBOX_NRFX_IPC
 config RPMSG_SERVICE_DUAL_IPM_SUPPORT
 	default RPMSG_SERVICE
 
+if IPC_SERVICE_BACKEND_RPMSG
+
+config IPC_SERVICE_BACKEND_RPMSG_SHMEM_RESET
+	default y
+
+endif # IPC_SERVICE_BACKEND_RPMSG
```
Just wondering, why this is defined in each board including nRF5340, instead of nrf5340 SoC?

kapi-no replied:
I observed that the configuration of other IPC modules (RPMsg service dual-mode) is defined in the boards Kconfig, so I assumed that there is some implicit agreement to have configuration there.

I can move this part to the SoC Kconfig to avoid duplications.

hubertmis replied:
I don't know this implicit agreement. Just curios why these duplications are needed. @carlocaione , do you know?

carlocaione replied:
IIRC this was due to a limitation of the legacy IPC services (like rpmsg service) but that should not be needed anymore with  IPC service so this can be moved to SoC level.

kapi-no replied:
I moved the configuration from boards Kconfig to SoC Kconfig. Please take a look

hubertmis reviewed (dismissed):
Looks good! Thanks

kapi-no committed:
soc: arm: nordic_nrf: enable shared memory reset for nRF53 App Core

Enabled the Shared Memory Reset feature for nRF5340 Application Core
when the RPMsg backend for IPC Service is enabled.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

carlocaione reviewed (dismissed):
LGTM

anangl commented on subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:
```diff
@@ -495,3 +495,25 @@ static int backend_init(const struct device *instance)
+#if defined(CONFIG_IPC_SERVICE_BACKEND_RPMSG_SHMEM_RESET)
+static int shared_memory_prepare(const struct device *arg)
+{
+	static struct backend_config_t
+		*config[DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT)];
+
+	DT_INST_FOREACH_STATUS_OKAY(BACKEND_CONFIG_INIT);
```
How about:
```suggestion
	static const struct backend_config_t *config[] = {
		DT_INST_FOREACH_STATUS_OKAY(BACKEND_CONFIG)
	};
```
with `#define BACKEND_CONFIG(n) &backend_config_##n,`?

kapi-no replied:
sure, this change will simplify implementation further

kapi-no replied:
fixed

hubertmis reviewed (dismissed):
Neat

carlocaione reviewed (approved):
`while(1) {
    LGTM
};`
